x-pack/legacy/plugins/apm/public/utils/__test__/flattenObject.test.ts

```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { flattenObject } from '../flattenObject';
+
+describe('FlattenObject', () => {
+  it('flattens multi level item', () => {
+    const data = {
+      foo: {
+        item1: 'value 1',
+        item2: { itemA: 'value 2' }
+      },
+      bar: {
+        item3: { itemA: { itemAB: 'value AB' } },
+        item4: 'value 4',
+        item5: [1],
+        item6: [1, 2, 3]
+      }
+    };
+
+    const flatten = flattenObject(data);
+    expect(flatten).toEqual([
+      { key: 'bar.item3.itemA.itemAB', value: 'value AB' },
+      { key: 'bar.item4', value: 'value 4' },
+      { key: 'bar.item5', value: 1 },
+      { key: 'bar.item6.0', value: 1 },
+      { key: 'bar.item6.1', value: 2 },
+      { key: 'bar.item6.2', value: 3 },
+      { key: 'foo.item1', value: 'value 1' },
+      { key: 'foo.item2.itemA', value: 'value 2' }
+    ]);
+  });
+  it('returns an empty array if no valid object is provided', () => {
+    expect(flattenObject({})).toEqual([]);
+  });
```

Comment:
What about this as well?
```ts
  it('returns an empty array if no valid object is provided', () => {
    expect(flattenObject({})).toEqual([]);
    expect(flattenObject(undefined)).toEqual([]);
    expect(flattenObject(null)).toEqual([]);
  });
```

Reply:
Typescript check will fail if undefined or null.

Reply:
Currently it's actually possible to call it like:
```ts
const rows = flattenObject(sectionProperties, section.key);
```

from `getMetadataItems`. I'd expect that to fail at runtime if `sectionProperties` is `null`. Therefore it's probably better to move the null check to `flattenObject`